File: 2024/q18.py
#!/usr/bin/env python
from collections import defaultdict, deque


def part1(data):
    grid = {(y,x):c for y,s in enumerate(data.strip().splitlines())
        for x,c in enumerate(s) if c in '.P'}
    palms = [p for p,c in grid.items() if c == 'P']
    start = next((y,x) for y,x in grid.keys() if x == 0)
    neib = [(1,0), (-1,0), (0,1), (0,-1)]

    fringe = deque([(0, start)])
    seen = set()
    goal = set(palms)
    while fringe:
        dist, pos = fringe.popleft()
        if pos in seen: continue
        seen.add(pos)
        if pos in goal:
            goal.remove(pos)
            if not goal:
                ans = dist
                break
        y,x = pos
        for dy,dx in neib:
            q = (y+dy, x+dx)
            if q in grid:
                fringe.append((dist+1, q))
    return ans


def part2(data):
    lines = data.strip().splitlines()
    w, h = len(lines[0]), len(lines)
    grid = {(y,x):c for y,s in enumerate(lines)
        for x,c in enumerate(s) if c in '.P'}
    palms = [p for p,c in grid.items() if c == 'P']
    start = [(y,x) for y,x in grid.keys() if x in [0,w-1]]
    neib = [(1,0), (-1,0), (0,1), (0,-1)]

    fringe = deque([(0, s) for s in start])
    seen = set()
    goal = set(palms)
    while fringe:
        dist, pos = fringe.popleft()
        if pos in seen: continue
        seen.add(pos)
        if pos in goal:
            goal.remove(pos)
            if not goal:
                ans = dist
                break
        y,x = pos
        for dy,dx in neib:
            q = (y+dy, x+dx)
            if q in grid:
                fringe.append((dist+1, q))
    return ans


def part3(data):
    lines = data.strip().splitlines()
    w, h = len(lines[0]), len(lines)
    grid = {(y,x):c for y,s in enumerate(lines)
        for x,c in enumerate(s) if c in '.P'}
    palms = [p for p,c in grid.items() if c == 'P']
    neib = [(1,0), (-1,0), (0,1), (0,-1)]

    def eval(start):
        fringe = deque([(0, start)])
        seen = set()
        goal = set(palms)
        res = 0
        while fringe:
            dist, pos = fringe.popleft()
            if pos in seen: continue
            seen.add(pos)
            if pos in goal:
                goal.remove(pos)
                res += dist
                if not goal:
                    return res
            y,x = pos
            for dy,dx in neib:
                q = (y+dy, x+dx)
                if q in grid:
                    fringe.append((dist+1, q))
    ans = float('inf')
    for p,c in grid.items():
        if c == '.':
            t = eval(p)
            ans = min(t, ans)
    return ans


def part3(data):
    lines = data.strip().splitlines()
    w, h = len(lines[0]), len(lines)
    grid = {(y,x):c for y,s in enumerate(lines)
        for x,c in enumerate(s) if c in '.P'}
    palms = [p for p,c in grid.items() if c == 'P']
    neib = [(1,0), (-1,0), (0,1), (0,-1)]
    acc = defaultdict(int)

    def eval(start):
        fringe = deque([(0, start)])
        seen = set()
        while fringe:
            dist, pos = fringe.popleft()
            if pos in seen: continue
            seen.add(pos)
            acc[pos] += dist
            y,x = pos
            for dy,dx in neib:
                q = (y+dy, x+dx)
                if q in grid:
                    fringe.append((dist+1, q))
    for p in palms:
        eval(p)
    ans = min(x for p,x in acc.items() if p not in palms)
    return ans


data = """
##########
..#......#
#.P.####P#
#.#...P#.#
##########
"""
assert part1(data) == 11

data = """
#######################
...P..P...#P....#.....#
#.#######.#.#.#.#####.#
#.....#...#P#.#..P....#
#.#####.#####.#########
#...P....P.P.P.....P#.#
#.#######.#####.#.#.#.#
#...#.....#P...P#.#....
#######################
"""
assert part2(data) == 21

data = """
##########
#.#......#
#.P.####P#
#.#...P#.#
##########
"""
assert part3(data) == 12


data = open('q18_p1.txt').read()
ans = part1(data)
print('part1:', ans)

data = open('q18_p2.txt').read()
ans = part2(data)
print('part2:', ans)

data = open('q18_p3.txt').read()
ans = part3(data)
print('part3:', ans)

File: event/2024/q18.jl
#!/usr/bin/env julia

const Pos = CartesianIndex{2}


function parsenotes(text)
    grid = stack(split(text), dims=1)
    P = findall(==('P'), grid)
    start = findfirst(==('.'), grid)
    return (grid, start, P)
end


function distance(a::Pos, b::Pos)
    sum(abs.((b - a).I))
end


function irrigate(grid, start, goal)
    neibs = [(1,0),(-1,0),(0,1),(0,-1)] .|> Pos
    fringe = Set(start)
    seen = Set{Pos}()
    for t in Iterators.countfrom(0)
        union!(seen, fringe)
        goal ⊆ seen && return t
        fringe = Set(q for p in fringe for q in p .+ neibs
            if q ∉ seen && get(grid, q, '#') != '#')
    end
end


function part1(data)
    grid, start, P = parsenotes(data)
    irrigate(grid, [start], P)
end


function part2(data)
    grid, start, P = parsenotes(data)
    start = [start, findlast(==('.'), grid)]
    irrigate(grid, start, P)
end


function part3(data)
    grid, _, P = parsenotes(data)
    neibs = [(1,0),(-1,0),(0,1),(0,-1)] .|> Pos
    acc = zeros(Int, size(grid))

    function irrigate(start)
        fringe = Set([start])
        seen = Set{Pos}()
        for t in Iterators.countfrom(0)
            acc[collect(fringe)] .+= t
            union!(seen, fringe)
            fringe = Set(q for p in fringe for q in p .+ neibs
                if q ∉ seen && get(grid, q, '#') != '#')
            isempty(fringe) && break
        end
    end

    irrigate.(P)
    ix = findall(==('.'), grid)
    minimum(acc[ix])
end


data = raw"
##########
..#......#
#.P.####P#
#.#...P#.#
##########
"
@assert part1(data) == 11


data = raw"
#######################
...P..P...#P....#.....#
#.#######.#.#.#.#####.#
#.....#...#P#.#..P....#
#.#####.#####.#########
#...P....P.P.P.....P#.#
#.#######.#####.#.#.#.#
#...#.....#P...P#.#....
#######################
"
@assert part2(data) == 21


data = raw"
##########
#.#......#
#.P.####P#
#.#...P#.#
##########
"
@assert part3(data) == 12


data = readchomp("q18_p1.txt")
println("part1: ", part1(data))
data = readchomp("q18_p2.txt")
println("part2: ", part2(data))
data = readchomp("q18_p3.txt")
println("part3: ", part3(data))
